NFIQ2/NFIQ2Algorithm/src/prediction/RandomForestML.cpp

```diff
@@ -123,6 +123,63 @@ NFIQ2::Prediction::RandomForestML::initModule(const std::string &fileName,
 	return hash;
 }
 
+#ifdef __ANDROID__
+std::string
+NFIQ2::Prediction::RandomForestML::initModule(AAssetManager *assets,
+    const std::string &fileName, const std::string &fileHash)
+{
+	if (assets == nullptr) {
+		throw NFIQ2::Exception(NFIQ2::ErrorCode::BadArguments,
+		    "Prediction::RandomForestML::initModule: AAssetManager is null");
+	}
+	if (fileName.empty()) {
+		throw NFIQ2::Exception(NFIQ2::ErrorCode::BadArguments,
+		    "Prediction::RandomForestML::initModule: "
+		    "file name is empty");
+	}
+	if (fileHash.empty()) {
+		throw NFIQ2::Exception(NFIQ2::ErrorCode::BadArguments,
+		    "Prediction::RandomForestML::initModule: "
+		    "hash value is empty");
+	}
+	AAsset *asset = AAssetManager_open(assets, fileName.c_str(),
+	    AASSET_MODE_STREAMING);
+	if (asset == nullptr) {
+		throw NFIQ2::Exception(NFIQ2::ErrorCode::InvalidConfiguration,
+		    "Prediction::RandomForestML::initModule: "
+		    "requested asset not found");
+	}
+	// reading the file with the asset manager
+	std::stringstream ss;
+	char buffer[BUFSIZ + 1]; // BUFSIZ defined in stdio.h
+	int cnt = AAsset_read(asset, buffer, BUFSIZ);
+	while (cnt > 0 && cnt <= BUFSIZ) {
+		buffer[cnt] = 0; // add terminating '\0'
+		ss << buffer;
```

Comment:
Are we guaranteed to not have a null-terminator anywhere in the model file? Maybe for the current model only? Should we `write` instead (if we keep this loop)?
```suggestion
		ss.write(buffer, cnt);
```

...or even store to a `std::vector<char>` instead?

Reply:
We read the forest params as string and the yaml file with the model is a text file. Using the '\0' is safe due it is the text termination and will not occur in the text file. 
I checked the reading routine in Android and 
- it does not need much time 
- the MD5 of the file content can successfully compared with the given hash string.
Moving this to a vector or messing with the string termination will easily invalidate the hash value.

Reply:
You can have `\0` in a text file and in the middle of a `string`. We might not in the current model, but using `operator<<` (vs `write` with a size) will not read all `BUFSIZ` bytes if a future model did have a `\0` inside. My overall question for this is and the two comments for line 154 are why are we reading in chunks and manipulating C-style strings if we don't have to with the API Android gives us?

Changing the storage container won't change the contents and we need to get to a C-style string for the digest function anyway (`vector<char>.data()`). I know `calculateHashString` takes a `string`, but it could just as easily and maybe more appropriately take something else in a refactor.

Reply:
@RLessmann: please ignore my second comment (we _have_ to get to a `std::string` to pass to OpenCV `cv::FileStorage` eventually, so the `vector` solution would duplicate RAM. However, I still find the `ss.write` to be more appropriate than `operator<<`. This should not change anything, avoids string conversion on every iteration, and guards us in the future if there is ever a `\0` in a file. Would you please make and test this change on device? It should be identical. I believe you can commit the suggestion below and run the branch.